### PR TITLE
Cnex array

### PR DESCRIPTION
**Pull request by gitlarryf**
Moved Cell.array to its own object.

### CODE REVIEW

File: exec/cnex/global.c [hunk 1]
```diff
@@ -89,15 +90,15 @@ void global_callFunction(const char *pszFunc, TExecutor *exec)
     fatal_error("global_callFunction(): \"%s\" - invalid or unsupported predefined function call.", pszFunc);
 }
 
-void print(TExecutor *exec)
+void print(struct tagTExecutor *exec)
```
Comment (ghewgill) [inline exec/cnex/global.c]: I don't understand why you keep changing these. What's the point of having the typedef to TExecutor if you're not going to use it? If you're going to use "struct tagTExecutor" everywhere, I'd rather see the typedef removed completely.

File: exec/cnex/array.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+typedef struct tagTArray {
+    size_t size;
+    struct tagTCell *data;
+} Array;
+
+Array *array_createArray(void);
+Array *array_createArrayFromSize(size_t iElements);
+
+void array_freeArray(Array *self);
+
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/array.h]: Currently, yes.  Perhaps I should put a future ToDo: comment in there?

File: exec/cnex/cell.c [hunk 3]
```diff
@@ -21,17 +22,15 @@ Cell *cell_fromAddress(Cell *c)
...
-    uint64_t i;
     Cell *a = cell_newCell();
 
     a->type = cArray;
-    a->array_size = c->array_size;
-    a->array = malloc(sizeof(Cell) * c->array_size);
-    if (a->array == NULL) {
-        fatal_error("Could not allocate memory for copy of array with %d elements.", c->array_size);
+    a->array = array_createArrayFromSize(c->array->size);
+    if (a->array->data == NULL) {
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/cell.c]: Removed from cell.c.

File: exec/cnex/cell.c [hunk 4]
```diff
@@ -95,22 +94,17 @@ Cell *cell_fromCell(const Cell *c)
...
-            x->array = malloc(sizeof(Cell) * c->array_size);
-            if (x->array == NULL) {
-                fatal_error("Could not allocate memory from array with %d elements.", c->array_size);
-            }
-            for (i = 0; i < x->array_size; i++) {
-                cell_resetCell(&x->array[i]);
-                cell_copyCell(&x->array[i], &c->array[i]);
+            x->array = array_createArrayFromSize(c->array->size);
+            for (size_t i = 0; i < x->array->size; i++) {
+                cell_resetCell(&x->array->data[i]);
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/cell.c]: Removed call to cell_resetCell().

File: exec/cnex/cell.c [hunk 5]
```diff
@@ -265,13 +248,11 @@ Cell *cell_arrayForWrite(Cell *c)
...
-        c->array = malloc(sizeof(Cell));
-        if (c->array == NULL) {
-            fatal_error("Could not allocate array for write.");
-        }
+    if (!c->array) {
+        c->array = array_createArrayFromSize(1);
     }
-    return c->array;
+    // ToDo: Validate return value!
+    return c->array->data;
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/cell.c]: I'm not sure about that!  Good catch.  Removed this function and the prototype.

File: exec/cnex/cell.c [hunk 6]
```diff
@@ -368,16 +348,12 @@ void cell_copyCell(Cell *dest, const Cell *source)
...
-        dest->array = malloc(sizeof(Cell) * dest->array_size);
-        if (dest->array == NULL) {
-            fatal_error("Could not allocate enough memory to copy array with %d elements.", dest->array_size);
-        }
-        for (i = 0; i < dest->array_size; i++) {
-            cell_resetCell(&dest->array[i]);
-            cell_copyCell(&dest->array[i], &source->array[i]);
+        dest->array = array_createArrayFromSize(source->array->size);
+        for (size_t i = 0; i < dest->array->size; i++) {
+            cell_resetCell(&dest->array->data[i]);
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/cell.c]: Removed call to cell_resetCel()

File: exec/cnex/global.c (hunk 1 above)
Comment (gitlarryf) [reply exec/cnex/global.c]: They are used by type everywhere in the code, except function calls.  This keeps my includes clean, so I don't have to include everything in every .h file.  I'd like to keep the typedefs, but at the same time, I can see removing them.  It just takes a lot more code to define them when I need them.

File: exec/cnex/array.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#ifndef _ARRAY_H
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/cnex/array.h]: Fixed, in every .h file in cnex.

**Pull request merged**